### PR TITLE
grpc-web sample

### PR DESCRIPTION
**Pull request by vmax**

### CODE REVIEW

File: app/client/BUILD [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  mode = "grpcweb"
+)
+
+
+closure_js_library(
+  name = "hello_world_service_js_library",
+  srcs = ["sample.js"],
+  deps = [":hello_world_service_grpc",
+          ":hello_world_service_closure_proto"],
+  suppress = ["strictMissingProperties", "reportUnknownTypes"]
```
Comment (oferb) [inline app/client/BUILD]: Why do we need to suppress these? Can you document the error?

File: app/client/BUILD [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+closure_proto_library(
+  name = "hello_world_service_closure_proto",
+  deps = ["//app/local_server/service:hello_world_service_proto"]
+)
+
+closure_grpc_web_library(
+  name = "hello_world_service_grpc",
+  deps = ["//app/local_server/service:hello_world_service_proto"],
+  mode = "grpcweb"
```
Comment (oferb) [inline app/client/BUILD]: As discussed, please link to https://github.com/grpc/grpc-web#wire-format-mode, and mention `grpcweb` and `grpcwebtext`

File: app/client/sample.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+
+/**
+ * Hello
+ */
+const hello = function () {
+  var client =  new HelloWorldServiceClient('http://localhost:8080', null, null);
+  var request = new HelloWorldRequest();
+  request.setMessage(document.getElementById("request-message").value);
+  client.helloWorld(request, {}, function(code, message) {
```
Comment (oferb) [inline app/client/sample.js]: Please document what `{}` is.

File: app/run_sample.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+cp bazel-bin/app/client/hello_world_service_js_binary.js app/client/app.js
+
+
+echo "Running local_server"
+bazel run //app/local_server:local_server &
+
+echo "Opening browser"
+$RUNNER $html_file
+
+echo "Running grpc web proxy"
```
Comment (oferb) [inline app/run_sample.sh:34]: Why not run the proxy before opening the browser?

**Review by oferb — CHANGES_REQUESTED** [review]

File: app/run_sample.sh (hunk 4 above)
Comment (vmax) [reply app/run_sample.sh:34]: @oferb running proxy is blocking whereas opening browser is not

Comment (oferb) [reply app/run_sample.sh:34]: Got it

File: app/client/BUILD [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  name = "hello_world_service_closure_proto",
+  deps = ["//app/local_server/service:hello_world_service_proto"]
+)
+
+
+# see https://github.com/grpc/grpc-web#wire-format-mode
+closure_grpc_web_library(
+  name = "hello_world_service_grpc",
+  deps = ["//app/local_server/service:hello_world_service_proto"],
+  mode = "grpcweb" # (grpcweb|grpcwebtext)
```
Comment (oferb) [inline app/client/BUILD:23]: `# mode can be grpcweb or grpcwebtext. For more details, see # see https://github.com/grpc/grpc-web#wire-format-mode`

**Review by oferb — APPROVED** [review]

**Comment by vmax** [comment]
@oferb please apply [0001-latest-fix-from-review.patch.gz](https://github.com/oferb/startup-os-example/files/2282307/0001-latest-fix-from-review.patch.gz) after merging:

```
gunzip 0001-latest-fix-from-review.patch.gz
git am -3 < 0001-latest-fix-from-review.patch
```

**Pull request merged**